src/riak_object.erl

```
@@ -239,30 +267,127 @@ compare_content_dates(C1,C2) ->
+                {true, true} ->
+                    %% Dot present on both sides, keep it
+                    Content = convert_to_dict_list(C0),
+                    MergeAcc#merge_acc{keep=ordsets:add_element(Content, Keep)};
+                {true, false} ->
+                    %% Dominated dot, add it to the drop list
+                    MergeAcc#merge_acc{drop=orddict:store(Dot, C0, Drop)};
+                {false, _} ->
+                    %% Not dominated, keep it
+                    Content = convert_to_dict_list(C0),
+                    MergeAcc#merge_acc{keep=ordsets:add_element(Content, Keep)}
+            end;
+        undefined ->
+            %% CRDTs and legacy data don't have dots, which are you?
+            case riak_kv_crdt:merge_value({Dict, Value}, {CRDT, [], []}) of
+                {CRDT, [], [E]} ->
+                    MergeAcc#merge_acc{error=[E | Error]};
+                {CRDT, [{Dict, Value}], []} ->
+                    Content = convert_to_dict_list(C0),
+                    MergeAcc#merge_acc{keep=ordsets:add_element(Content, Keep)};
+                {CRDT2, [], []} ->
+                    MergeAcc#merge_acc{crdt=CRDT2}
+            end
+    end.
+
+%% @private Transform a `merge_acc()' to a list of `r_content()'
+-spec merge_acc_to_contents(merge_acc()) -> list(r_content()).
+merge_acc_to_contents(MergeAcc) ->
+    #merge_acc{keep=Keep, crdt=CRDTs} = MergeAcc,
+    Keep2 = lists:map(fun convert_from_dict_list/1, ordsets:to_list(Keep)),
+    orddict:fold(fun(_Type, {Meta, CRDT}, Acc) ->
+                         [{r_content, riak_kv_crdt:meta(Meta, CRDT),
+                           riak_kv_crdt:to_binary(CRDT)} | Acc]
+                 end,
+                 Keep2,
+                 CRDTs).
+
+%% @private Get the dot from the passed metadata dict
+%% (if present and valid).
+-spec get_dot(dict()) -> {ok, vclock:dot()} | undefined.
+get_dot(Dict) ->
```

None that I can think of. Do you think we should a) log here or b) just let it crash at the point the dot is compared to the clock? c) crash here?